### PR TITLE
Refactor & fix integration tests on RBAC branch

### PR DESCRIPTION
**Pull request by Dostonbek1**
This PR aims to refactor existing integration tests as well as updating them to adjust to the new RBAC implementation, although the bulk of the changes are refactors for tests to use fixtures. 

EDIT:
All tests are now passing 🟢

### CODE REVIEW

File: tests/integration/api/test_activation.py [hunk 1]
```diff
@@ -362,51 +362,32 @@ def test_disable_activation(
...
         f"{api_url_v1}/activations/{default_activation.id}/instances/"
     )
     data = response.data["results"]
     assert response.status_code == status.HTTP_200_OK
     assert len(data) == len(instances)
-    assert data[0]["name"] == instances[0].name
-    assert data[1]["name"] == instances[1].name
+    # invert the indices because response data is descending by default
+    assert data[0]["name"] == instances[1].name
+    assert data[1]["name"] == instances[0].name
```
Comment (jshimkus-rh) [inline tests/integration/api/test_activation.py:378]: Performing a suitable sort on both lists will insulate this test from any future changes that affect response content ordering.

Comment (Dostonbek1) [reply tests/integration/api/test_activation.py:378]: Good point. I will add sorting for both lists in the test.

**Comment by Dostonbek1** [comment]
Update:
```
104 failed, 345 passed, 28 skipped, 3 warnings
```

File: tests/integration/api/conftest.py [hunk 2]
```diff
@@ -270,6 +272,56 @@ def new_activation(
     )
 
 
+@pytest.fixture
+def default_activation_instances(
+    default_activation: models.Activation, default_project: models.Project
+) -> models.RulebookProcess:
+    """Return a list of Activation Instances"""
+    return models.RulebookProcess.objects.bulk_create(
```
Comment (Alex-Izquierdo) [inline tests/integration/api/conftest.py:280]: I think one of our problems is that we rely on models instead of objects created by the api. Through the model, we skip the logic in the serializer and with bulk we skip logic in the save method. Have you considered to create the objects through api requests?

**Review by Alex-Izquierdo — COMMENTED** [review]
Is the plan to unify fixture across all the tests or only for the api? because there are duplicated or similar fixtures in other places that can be also reused. 
For example:
https://github.com/ansible/eda-server/blob/main/tests/integration/services/activation/test_activation.py
https://github.com/ansible/eda-server/blob/main/tests/integration/services/activation/test_activation.py


Anycase, @Dostonbek1 thank you very much for this effort. It is a good improvement. 

There is a internal ticket that I created to approach some of this actions. You may want to reuse it: https://issues.redhat.com/browse/AAP-15713

**Comment by Dostonbek1** [comment]
> Is the plan to unify fixture across all the tests or only for the api? because there are duplicated or similar fixtures in other places that can be also reused. For example: https://github.com/ansible/eda-server/blob/main/tests/integration/services/activation/test_activation.py https://github.com/ansible/eda-server/blob/main/tests/integration/services/activation/test_activation.py
> 
> Anycase, @Dostonbek1 thank you very much for this effort. It is a good improvement.
> 
> There is a internal ticket that I created to approach some of this actions. You may want to reuse it: https://issues.redhat.com/browse/AAP-15713

@Alex-Izquierdo I have been going through tests file by file in unit tests and have only done test_activation and test_activation_instance so far. That's why you might see the rest of the tests are still waiting to be refactored. But the goal is to make fixtures to be reusable across all tests and avoid duplication as much as possible.

File: tests/integration/api/conftest.py (hunk 2 above)
Comment (Dostonbek1) [reply tests/integration/api/conftest.py:280]: I agree. But I actually don't have much info on what the standard conventions are regarding using fixtures for object creations in tests. I know we already do what you are suggesting in e2e tests because that's the only way in e2e tests. Anyway, I'm open to going that way if others also deem it more reasonable.

Comment (jshimkus-rh) [reply tests/integration/api/conftest.py:280]: As with many things, it depends. 

1. If a test is exercising a class behavior it's reasonable to use the class directly avoiding potential issues from other source.
2. If a test is exercising a higher-level behavior that requires an instance of a particular class creating the instance at the level of that behavior (assuming that is where it is normally created) is reasonable by the same rationale as in #1.  
3. Any other creation of an instance would be done as in production, as it would be assumed to work, providing the benefit of early detection of issues.

Of course, the above isn't meant to be a straitjacket.  If there are issues with code that is not under test getting in the way of progress relaxing the above may be the correct thing to do.  Though it should be noted so that it can later be "corrected."

Comment (jshimkus-rh) [reply tests/integration/api/conftest.py:280]: Making changes to adhere to what I described would not be a short-term effort. 🤕

**Comment by Dostonbek1** [comment]
Update after refactoring test_credential tests:
```
97 failed, 352 passed, 28 skipped, 3 warnings
```

File: tests/integration/api/conftest.py (hunk 2 above)
Comment (Alex-Izquierdo) [reply tests/integration/api/conftest.py:317]: Agree with all. @Dostonbek1 It should not be a effort to address everything. We can split it and iterate. But I think it can be a good opportunity to define the foundations. I think we can make https://issues.redhat.com/browse/AAP-15713 a epic so we can create multiple tickets there. 

Regarding models I would say that by default, we always want the object created through the request. There may be exceptions, but I think there will be less expections in that way rather than models.

Comment (Dostonbek1) [reply tests/integration/api/conftest.py:516]: @jshimkus-rh @Alex-Izquierdo Thank you both for feedback/thoughts. I have reached a point where I refactored most of the test modules to use fixtures and all tests are now passing. In some of the modules, I made changes just enough for tests to pass and left comments as needed for future refactoring; I wanted to avoid making this PR bigger than it already is and we can work on those test modules in future iterations. Please, give this PR another look. :)

**Comment by Dostonbek1** [comment]
All tests are passing now 🥳

File: src/aap_eda/api/views/user.py [hunk 3]
```diff
@@ -156,15 +154,7 @@ def get_queryset(self):
...
-        try:
-            serializer.save(user=self.request.user)
-        except django.db.utils.IntegrityError:
-            name_exists = models.AwxToken.objects.filter(
-                user=self.request.user, name=serializer.validated_data["name"]
-            ).exists()
-            if name_exists:
-                raise Conflict("Token with this name already exists.")
-            raise Conflict
+        serializer.save(user=self.request.user)
```
Comment (AlanCoding) [inline src/aap_eda/api/views/user.py:157]: Sanity check of diff - `perform_create` override here is still necessary because the intent is to get the user from the _request_, not from the _request data_.

File: tests/integration/dab_rbac/conftest.py [hunk 4]
```diff
@@ -64,6 +64,10 @@ def make_name(self):
 
     def fk_fields(self, cls):
         for field in cls._meta.concrete_fields:
+            # corner case for when either an activation or a event_stream
+            # must be set, but not both
+            if cls == RulebookProcess and field.name == "event_stream":
+                continue
```
Comment (AlanCoding) [inline tests/integration/dab_rbac/conftest.py:70]: Nice to see this worked :D

File: tests/integration/api/conftest.py [hunk 5]
```diff
@@ -41,13 +40,12 @@ def admin_user(default_organization):
...
     )
-    organization = models.Organization.objects.get_default()
     admin_role = RoleDefinition.objects.create(
         name="Test Admin",
-        content_type=ContentType.objects.get_for_model(organization),
+        content_type=ContentType.objects.get_for_model(default_organization),
     )
     admin_role.permissions.add(*DABPermission.objects.all())
-    admin_role.give_permission(user, organization)
+    admin_role.give_permission(user, default_organization)
```
Comment (AlanCoding) [inline tests/integration/api/conftest.py:48]: :+1: all looks good here. I know this is convoluted, but this did help clear up some issues.

So the testing system is roughly "same as the old system" but in the default organization. Except user permissions and some other corner cases...

File: src/aap_eda/api/views/user.py (hunk 3 above)
Comment (Dostonbek1) [reply src/aap_eda/api/views/user.py:157]: Correct. This POST endpoint doesn't accept `user` data but rather gets the user making the request to assign in the AxwToken object. This makes sense because users can only create tokens for themselves, not for other users.

**Review by AlanCoding — APPROVED** [review]

**Review by jshimkus-rh — APPROVED** [review]
